### PR TITLE
Nav2 route tool

### LINKED ISSUE #5: Nav2 Route Tool
### Are you Saxton Lab/Leidos employee?

Yes

### Summary

Manually creating GeoJSON files that integrate with the [Navigation2 Route Server](https://github.com/ros-navigation/navigation2/tree/nav2_route_server/nav2_route) is a tedious and time-intensive process. An Rviz plugin is needed to help speed up the process of creating and editing these files. 

### Reasoning for new functionality

_No response_

### PR DESCRIPTION
**Pull request by john-chrosniak**
# PR Details
## Description

This PR introduces a tool for creating and editing geojson route graphs used in the Nav2 Route Server.

## Related GitHub Issue

Closes #5 

## Related Jira Key

[CF-867](https://usdot-carma.atlassian.net/browse/CF-867)

## Motivation and Context

This tool was developed to speed up the creation and editing of route graphs. They originally were very tedious to make and edit.

## How Has This Been Tested?

Testing has included creating and editing a variety of route graphs. Additionally, route graphs with various operations and embedded metadata have been loaded and saved to ensure that no fields are erased by the graph saver.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

[CF-867]: https://usdot-carma.atlassian.net/browse/CF-867?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: nav2_route/include/nav2_route/graph_saver.hpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * @class nav2_route::GraphSaver
+ * @brief An action server to save graph objects to a file
+ */
+class GraphSaver
+{
+public:
+  /**
+   * @brief A constructor for nav2_route::GraphSaver
+   * @param options Additional options to control creation of the node.
```
Comment (adev4a) [inline nav2_route/include/nav2_route/graph_saver.hpp]: Seems like the constructor here is taking different arguments, can we update this text?

File: nav2_route/include/nav2_route/graph_saver.hpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2024, Leidos
```
Comment (adev4a) [inline nav2_route/include/nav2_route/graph_saver.hpp]: Can we remove the "," after 2024 to be consistent with the rest of our code?

File: nav2_route/include/nav2_route/graph_saver.hpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+   */
+  bool transformGraph(Graph & graph);
+
+  std::string route_frame_, graph_filepath_;
+  std::shared_ptr<tf2_ros::Buffer> tf_;
+  rclcpp::Logger logger_{rclcpp::get_logger("GraphSaver")};
+
+  // Graph Parser
+  pluginlib::ClassLoader<GraphFileSaver> plugin_loader_;
+  GraphFileSaver::Ptr graph_file_saver_;
```
Comment (adev4a) [inline nav2_route/include/nav2_route/graph_saver.hpp:84]: This is a little confusing to read without checking the GraphFileSaver definition to see what ::Ptr means. Can we just use std::shared_ptr here?

File: nav2_route/src/graph_saver.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2024, Leidos
```
Comment (adev4a) [inline nav2_route/src/graph_saver.cpp]: Lets remove the "," after 2024

File: nav2_route/include/nav2_route/interfaces/graph_file_saver.hpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+   * @brief Configure the graph file saver, but do not store the node
+   * @param parent pointer to user's node
+   */
+  virtual void configure(
+    const rclcpp_lifecycle::LifecycleNode::SharedPtr node) = 0;
+
+  /**
+   * @brief Method to save the graph to the filepath
+   * @param graph The graph to save
+   * @param filepath The file to save to
```
Comment (adev4a) [inline nav2_route/include/nav2_route/interfaces/graph_file_saver.hpp]: The "path" to save the file to?

File: nav2_route/include/nav2_route/plugins/graph_file_savers/geojson_graph_file_saver.hpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+   */
+  bool saveGraphToFile(
+    Graph & graph,
+    std::string filepath) override;
+
+protected:
+  /**
+   * @brief Add nodes into the graph
+   * @param[out] graph The graph that will contain the new nodes
+   * @param[in] nodes The nodes to be added into the graph
```
Comment (adev4a) [inline nav2_route/include/nav2_route/plugins/graph_file_savers/geojson_graph_file_saver.hpp]: Is "nodes" in the parameter in here supposed to be json_features?

File: nav2_route/src/graph_saver.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  auto graph_file_saver_id = node->get_parameter("graph_file_saver").as_string();
+  if (graph_file_saver_id == default_plugin_id_) {
+    nav2_util::declare_parameter_if_not_declared(
+      node, default_plugin_id_ + ".plugin", rclcpp::ParameterValue(default_plugin_type));
+  }
+
+  // Create graph file saver plugin
+  try {
+    plugin_type_ = nav2_util::get_plugin_type_param(node, graph_file_saver_id);
+    GraphFileSaver::Ptr graph_parser = plugin_loader_.createSharedInstance((plugin_type_));
```
Comment (adev4a) [inline nav2_route/src/graph_saver.cpp]: This intermediate pointer might not be needed. graph_file_saver can directly be assigned here

File: nav2_route/src/graph_saver.cpp [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  Graph & graph,
+  std::string filepath)
+{
+  if (filepath.empty() && !graph_filepath_.empty()) {
+    RCLCPP_DEBUG(
+      logger_, "The graph filepath was not provided. "
+      "Setting to %s", graph_filepath_.c_str());
+    filepath = graph_filepath_;
+  } else if (filepath.empty() && graph_filepath_.empty()) {
+    // No graph to try to save
```
Comment (adev4a) [inline nav2_route/src/graph_saver.cpp:74]: Can we log a warning here about no filepath being defined?

File: nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  const rclcpp_lifecycle::LifecycleNode::SharedPtr node)
+{
+  RCLCPP_INFO(node->get_logger(), "Configuring geojson graph file saver");
+  logger_ = node->get_logger();
+}
+
+bool GeoJsonGraphFileSaver::saveGraphToFile(
+  Graph & graph, std::string filepath)
+{
+  Json json_graph, json_crs, json_properties;
```
Comment (adev4a) [inline nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp:35]: Can we put this in a try catch? streaming data from a file is likely to run into issues.

File: nav2_route/src/graph_saver.cpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    return false;
+  }
+
+  RCLCPP_INFO(
+    logger_,
+    "Saving graph file from %s, by parser %s", filepath.c_str(), plugin_type_.c_str());
+
+  if (!graph_file_saver_->saveGraphToFile(graph, filepath)) {
+    return false;
```
Comment (adev4a) [inline nav2_route/src/graph_saver.cpp:87]: Lets add a warning here or in the function being called indicating why this failed

File: nav2_route/src/graph_saver.cpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+  return true;
+}
+
+bool GraphSaver::transformGraph(Graph & graph)
+{
+  std::unordered_map<std::string, tf2::Transform> cached_transforms;
+  for (auto & node : graph) {
+    std::string node_frame = node.coords.frame_id;
+    if (node_frame.empty() || node_frame == route_frame_) {
+      continue;
```
Comment (adev4a) [inline nav2_route/src/graph_saver.cpp:108]: Can we add a comment here explaining why we ignore this condition? It is obvious in this case when we look at the transform calculation below, but adding guiding comments generally make reading the code easier.

File: nav2_route/src/graph_saver.cpp [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    if (cached_transforms.find(node_frame) == cached_transforms.end()) {
+      tf2::Transform tf_transform;
+      bool got_transform = nav2_util::getTransform(
+        node_frame, route_frame_, tf2::durationFromSec(0.1), tf_, tf_transform);
+
+      if (!got_transform) {
+        return false;
```
Comment (adev4a) [inline nav2_route/src/graph_saver.cpp:120]: Would it be meaningful here to return a warning about which node frame was causing an error? Or are node_frames hard for a user to interpret directly?

File: nav2_route/src/graph_saver.cpp [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    tf2::Vector3 graph_coord(
+      node.coords.x,
+      node.coords.y,
+      0.0);
+
+    tf2::Vector3 new_coord = cached_transforms[node_frame] * graph_coord;
+
+    node.coords.x = static_cast<float>(new_coord.x());
```
Comment (adev4a) [inline nav2_route/src/graph_saver.cpp:133]: Lets add some comments to explain what's happening here. It seems like we are looping through the graph coordinates and updating them to route coordinates. Is that correct?

Comment (adev4a) [reply nav2_route/src/graph_saver.cpp:133]: nit: Can we update the function name to indicate transform what to what?

File: nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+#include <string>
+#include <vector>
+#include <fstream>
+
+#include "nav2_route/plugins/graph_file_savers/geojson_graph_file_saver.hpp"
+
+namespace nav2_route
+{
+
+void GeoJsonGraphFileSaver::configure(
```
Comment (adev4a) [inline nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp:25]: nit: This function shares its name with a ros2 lifecycle function used to configure a node. It might be clearer to rename this function to something like initialize/configure_logger.

File: nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+      json_edges.push_back(json_edge);
+    }
+  }
+}
+
+void GeoJsonGraphFileSaver::convertMetaDataToJson(
+  const Metadata & metadata, Json & json_metadata) 
+{
+  for (auto itr = metadata.data.begin(); itr != metadata.data.end(); itr++) {
+    if (itr->second.type() == typeid(std::string)) {
```
Comment (adev4a) [inline nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp:117]: In this function use Metadata getValue() instead https://github.com/usdot-fhwa-stol/navigation2/blob/9e076076500b1e75fd4d5511e1d0ad9b9dfda8b4/nav2_route/include/nav2_route/types.hpp#L40

File: nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp [hunk 16]
```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2024, Leidos
```
Comment (adev4a) [inline nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp]: Remove the ","

File: nav2_rviz_plugins/launch/route_tool.launch.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+        output='screen',
+        parameters=[{'yaml_filename' : map_file}]
+    )
+
+    start_lifecycle_manager_cmd = launch_ros.actions.Node(
+        package='nav2_lifecycle_manager',
+        executable='lifecycle_manager',
+        name='lifecycle_manager',
+        output='screen',
+        emulate_tty=True,  # https://github.com/ros2/launch/issues/188
```
Comment (adev4a) [inline nav2_rviz_plugins/launch/route_tool.launch.py]: Can we add a short description for what this issue means for us here?

File: nav2_rviz_plugins/src/route_tool.cpp [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+        graph_vis_publisher_ = node_->create_publisher<visualization_msgs::msg::MarkerArray>(
+            "route_graph", rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable());
+        node_->activate();
+        tf_ = std::make_shared<tf2_ros::Buffer>(node_->get_clock());
+        graph_loader_ = std::make_shared<nav2_route::GraphLoader>(node_, tf_, "map");
+        graph_saver_ = std::make_shared<nav2_route::GraphSaver>(node_, tf_, "map");
+        ui_->add_node_button->setChecked(true);
+        ui_->edit_node_button->setChecked(true);
+        ui_->remove_node_button->setChecked(true);
+        graph_.reserve(10000);
```
Comment (adev4a) [inline nav2_rviz_plugins/src/route_tool.cpp]: nit: Why is this required? Can we add a comment for why this number was picked?

File: nav2_rviz_plugins/launch/route_tool.launch.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2024, Leidos
```
Comment (adev4a) [inline nav2_rviz_plugins/launch/route_tool.launch.py]: Remove the ","

File: nav2_rviz_plugins/src/route_tool.cpp [hunk 20]
```diff
@@ -0,0 +1,274 @@
+// Copyright (c) 2024, Leidos
```
Comment (adev4a) [inline nav2_rviz_plugins/src/route_tool.cpp]: Remove the ","

File: nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+            void on_edit_node_button_toggled(void);
+
+        /**
+         *  Finally, we close up with protected member variables
+         */
+        protected:
+            // UI pointer
+            std::unique_ptr<Ui::route_tool> ui_;
+            // ROS declaration
```
Comment (adev4a) [inline nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp]: This comment can be removed or moved to where the ros declaration is happening now

File: nav2_route/include/nav2_route/graph_saver.hpp (hunk 3 above)
Comment (john-chrosniak) [reply nav2_route/include/nav2_route/graph_saver.hpp:84]: I agree that would be more clear but I wanted to keep this class similar to the graph loader: https://github.com/usdot-fhwa-stol/navigation2/blob/route-server-develop/nav2_route/include/nav2_route/graph_loader.hpp#L87

**Comment by adev4a** [comment]
Since this change is creating functionality that is intended to be merged into nav2, i would also recommend creating a Github issue detailing the motivation behind the change

File: nav2_route/src/graph_saver.cpp (hunk 13 above)
Comment (john-chrosniak) [reply nav2_route/src/graph_saver.cpp:133]: I added comments but think the function name should stay the same to match the graph loader: https://github.com/usdot-fhwa-stol/navigation2/blob/route-server-develop/nav2_route/src/graph_loader.cpp#L98

File: nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp (hunk 14 above)
Comment (john-chrosniak) [reply nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp:25]: This is intentional since it is overriding the lifecycle node's implementation.

File: nav2_rviz_plugins/launch/route_tool.launch.py (hunk 17 above)
Comment (john-chrosniak) [reply nav2_rviz_plugins/launch/route_tool.launch.py]: This actually wasn't needed so I removed it

File: nav2_rviz_plugins/src/route_tool.cpp (hunk 18 above)
Comment (john-chrosniak) [reply nav2_rviz_plugins/src/route_tool.cpp]: This was needed for a prior implementation but isn't anymore, so removing this

File: nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp (hunk 15 above)
Comment (john-chrosniak) [reply nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp:117]: I'm not sure this would help with cleanup too much since this would change the logic to iterate through just map keys and then call `getValue` to get the value, whereas I'm currently iterating through key/value pairs.

File: nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+  }
+}
+
+void GeoJsonGraphFileSaver::convertOperationsToJson(
+  const Operations & operations, Json & json_operations)
+{
+  for (const auto & operation : operations) {
+    Json json_operation, json_metadata;
+    json_operation["type"] = operation.type;
+    json_operation["trigger"] = operation.trigger;  // TODO: Needs to convert int to string
```
Comment (willjohnsonk) [inline nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp]: I see a TODO is still marked here, do you want to address this or is it no rush?

File: nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp [hunk 23]
```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2024 Leidos
```
Comment (willjohnsonk) [inline nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp]: this is peak nitpick but (c) should be uppercase (C) to represent the Copyright symbol. I think this needs to be updated in most of the files

File: nav2_rviz_plugins/resource/route_tool.ui
Comment (willjohnsonk) [inline nav2_rviz_plugins/resource/route_tool.ui:1]: xmls should also have copyright information, you can copy the ones from other CARMA packages. I'd make sure each new file has the appropriate copyright information since industry will potentially be using our work

**Review by willjohnsonk — CHANGES_REQUESTED** [review]
Mostly nitpicks related to ownership information

File: nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp (hunk 22 above)
Comment (john-chrosniak) [reply nav2_route/src/plugins/graph_file_savers/geojson_graph_file_saver.cpp]: TODO was actually not needed so removed the comment

File: nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp (hunk 23 above)
Comment (john-chrosniak) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp]: It looks like the rest of nav2 uses a lowercase c, so i left this for the sake of consistency

File: nav2_rviz_plugins/resource/route_tool.ui
Comment (john-chrosniak) [reply nav2_rviz_plugins/resource/route_tool.ui:1]: added!

File: nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp (hunk 23 above)
Comment (willjohnsonk) [reply nav2_rviz_plugins/include/nav2_rviz_plugins/route_tool.hpp]: 100% should be a capital, time to create an official Nav2 issue for improper use of Copyright as a capitalized label (your choice but this is a standard `It is designed as an encircled capital C © to refer to the “copyright.”` )

**Review by adev4a — APPROVED** [review]

**Review by willjohnsonk — APPROVED** [review]
